master/docs/developer/pull-request.rst

```diff
@@ -0,0 +1,212 @@
+
+* Clone your forked repository on your local machine, so you can do your hacking.
+  GitHub will display a link titled "Your Clone URL".
+  Click this link to see instructions for cloning your URL.
+  It's something like:
+
+.. code-block:: bash
+
+    git clone git@github.com:myusername/buildbot.git
+    cd buildbot
+
+* Locally, create a new branch based on the `master` branch:
+
+.. code-block:: bash
+
+    git checkout -b myfixes origin/master
+
+* Hack mercilessly.
+  If you're a git afficionado, you can make a neat and pretty commit sequence; otherwise, just get it done.
+  Don't forget to add new test cases and any necessary documentation.
+
+* Test your changes.
+  See :ref:`LocalTestingCheatSheet` for instructions of how to launch various tests locally.
+
+* Commit.
+  For this step it's best to use a GUI for Git.
+  See this `list <https://git-scm.com/downloads/guis>`_ of known Git GUIs.
+  If you only want to use the shell, do the following:
+
+.. code-block:: bash
+
+    git add $files_that_matter
+    git commit
+
+* When you're confident that everything is as it should be, push your changes back to your repository on GitHub, effectively making them public.
+
+.. code-block:: bash
+
+    git push origin myfixes
+
+* Now all that's left is to let the Buildbot developers know that you have patches awaiting their attention.
+  In your web browser, go to your repository (you may have to hit "reload") and choose your new branch from the "all branches" menu.
+
+* Double-check that you're on your branch, and not on a particular commit.
+  The current URL should end in the name of your patch, not in a SHA1 hash.
+
+* Click “Pull Request”
+
+* Double-check that the base branch is "buildbot/buildbot@master".
+  If your repository is a fork of the buildbot/buildbot repository, this should already be the case.
+
+* Fill out the details and send away!
+
+.. _LocalTestingCheatSheet:
+
+Local testing cheat sheet
+-------------------------
+
+This section details how to locally run the test suites that are run by BuildBot during each PR.
+Not all test suites have been documented so far, only these that fail most often.
+Before each of the commands detailed below, a virtualenv must be setup as described in :ref:`PythonDevQuickStart`:
+
+.. code-block:: bash
+
+    make virtualenv
+    . .venv/bin/activate
+
+If you see weird test results after changing branches of the repository, remove the `.venv` directory and repeat above again.
+Note that `pip install -r <file>.txt` only need to be run once at the beginning of your testing session.
+
+Master unit tests
+~~~~~~~~~~~~~~~~~
+
+Tests in this category run the Python unit tests for the master.
+These tests are represented by **bb/trial/** test names in the BuildBot CI.
+To run locally, execute the following:
+
+.. code-block:: bash
+
+    pip install -r requirements-ci.txt
+    trial -j16 buildbot  # change -j parameter to fit the number of cores you have
```

Comment:
We should use a more conservative `-j` value because many will run it verbatim.

Reply:
Replaced with -j8. Even the weakest machines now seem to have 2 cores with 4 hyperthreads, developer machines are likely to have 4 cores with 8 hyperthreads.